src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,225 @@
+                context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
+                    context.Compilation))
+            {
+                return;
+            }
+
+            var typesUsedPerParameter = PooledConcurrentDictionary<IParameterSymbol, PooledConcurrentSet<ITypeSymbol>>.GetInstance(SymbolEqualityComparer.Default);
+            foreach (var param in methodSymbol.Parameters)
+            {
+                typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                var paramReference = (IParameterReferenceOperation)context.Operation;
+
+                switch (paramReference.Parent.Kind)
+                {
+                    case OperationKind.Invocation:
+                        AddDeclaringTypeSymbol(((IInvocationOperation)paramReference.Parent).TargetMethod);
+                        break;
+
+                    case OperationKind.PropertyReference:
+                    case OperationKind.EventReference:
+                    case OperationKind.FieldReference:
+                    case OperationKind.MethodReference:
+                        AddDeclaringTypeSymbol(((IMemberReferenceOperation)paramReference.Parent).Member);
+                        break;
+
+                    case OperationKind.Argument:
+                        AddType(paramReference.Type);
+                        break;
+
+                    case OperationKind.Conversion:
+                        AddType(paramReference.Parent.Type);
+                        break;
+
+                    default:
+                        break; // Do nothing
+                }
+
+                // Local functions
+                void AddType(ITypeSymbol typeSymbol)
+                {
+                    if (typesUsedPerParameter.TryGetValue(paramReference.Parameter, out var typePool))
+                    {
+                        typePool.Add(typeSymbol);
+                    }
+                }
+
+                void AddDeclaringTypeSymbol(ISymbol symbol)
+                    => AddType(FindDeclaringType(symbol));
+            }, OperationKind.ParameterReference);
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                try
+                {
+                    var overloads = methodSymbol.GetOverloads();
+
+                    foreach (var (parameter, usedTypesPool) in typesUsedPerParameter)
+                    {
+                        var usedTypes = usedTypesPool.ToImmutableArray();
+                        // If type is unused, we ignore it. There is another rule to detect
+                        // unused parameters.
+
+                        if (usedTypes.Length == 1)
+                        {
+                            ReportIfTypesAreDifferent(parameter, usedTypes[0], overloads);
```

Comment:
What if the used type is even less generic?

```cs
// Base => Derived => FurtherDerived
void M(Derived d)
{
   d.MethodOnDerived();
   if (...) // is always FurtherDerived?
   {
      ((FurtherDerived)d).MethodOnFurtherDerived();
   }  
}
```

I think we should not suggest changing the parameter type to `FurtherDerived`

Reply:
I am not sure to follow. Do you mean we should not suggest to replace `Derived` by `Base` in the method signature? For me it would makes sense to still suggest so.

Reply:
> Do you mean we should not suggest to replace Derived by Base

I meant we should not suggest changing `Derived` to `FurtherDerived` here. I don't think anything inside `ReportIfTypesAreDifferent` checks if the suggested type is a more generic or less generic type and we skip invoking `FindMostGenericType` when `usedTypes.Length == 1`